kore/src/Kore/Syntax/Json/Internal.hs

```diff
@@ -0,0 +1,687 @@
+ Unicode escape sequences with 2, 4 or 8 hex: \xHH, \uHHHH, \UHHHHHHHH
+-}
+checkStringChars :: Text -> [String]
+checkStringChars txt =
+    [ "Contains non-printable characters: " <> (show . nub $ T.unpack nonPrintables)
+    | not $ T.null nonPrintables
+    ]
+        ++ ["Contains invalid escape sequences: " <> showList badEscapes "" | not $ null badEscapes]
+  where
+    nonPrintables = T.filter (not . isPrint) txt
+    badEscapes = reverse . fst $ T.foldl' collectEscapes ([], Normal) txt
+
+-- Text fold function to collect invalid escape sequences in a string literal
+data CollectState
+    = Normal
+    | Escape
+    | UTF Int String
+    | Bad Int String
+    deriving stock (Show)
+
+collectEscapes :: ([String], CollectState) -> Char -> ([String], CollectState)
+-- normal mode
+collectEscapes (acc, Normal) '\\' = (acc, Escape) -- enter escaped mode
+collectEscapes (acc, Normal) _ = (acc, Normal) -- could check if printable here
+
+-- escaping mode
+collectEscapes (acc, Escape) 'x' = (acc, UTF 2 "x\\") -- enter UTF mode, expect 2 hex
+collectEscapes (acc, Escape) 'u' = (acc, UTF 4 "u\\") -- enter UTF mode, expect 4 hex
+collectEscapes (acc, Escape) 'U' = (acc, UTF 8 "U\\") -- enter UTF mode, expect 8 hex
+collectEscapes (acc, Escape) c
+    | c `elem` ['t', 'n', 'f', 'r', '\"', '\\'] -- good escape
+        =
+        (acc, Normal)
+    | otherwise -- bad escape
+        =
+        (['\\', c] : acc, Normal)
+-- UTF escape mode (n = remaining hex digits expected)
+collectEscapes s@(acc, UTF n part) c
+    | n > 8 || n < 0 = error $ "Illegal collect state" <> show s
+    | n == 0 = collectEscapes (acc, Normal) c -- good UTF escape
+    | c `elem` hexDigits = (acc, UTF (n -1) (c : part))
+    | otherwise = (acc, Bad (n - 1) (c : part))
+  where
+    hexDigits = ['0' .. '9'] <> ['A' .. 'F'] <> ['a' .. 'f']
+
+-- already bad, collect remaining expected hex digits no matter what they are)
+collectEscapes s@(acc, Bad n part) c
+    | n > 7 || n < 1 = error $ "Illegal collect state" <> show s
+    | n == 1 = (reverse part : acc, Normal)
+    | otherwise = (acc, Bad (n - 1) (c : part))
+
+------------------------------------------------------------
+data Sort
+    = Sort
+        { name :: Id
+        , args :: [Sort]
+        }
+    | SortVariable
+        { name :: Id
+        }
+    deriving stock (Eq, Show, Generic)
+
+instance ToJSON Sort where
+    toJSON = genericToJSON codecOptions
+
+instance FromJSON Sort where
+    parseJSON = genericParseJSON codecOptions
+
+data LeftRight
+    = Left
+    | Right
+    deriving stock (Eq, Show, Generic, Enum, Bounded)
+    deriving anyclass (ToJSON, FromJSON)
+
+------------------------------------------------------------
+-- reading
+
+-- | Errors relating to the json codec, re-exported from API
+data JsonError
+    = -- | Problem reported by json parser
+      ParseError String
+    | -- | MultiOr/MultiApp require a non-empty argument list
```

Comment:
Maybe worth using [NonEmpty](https://hackage.haskell.org/package/base-4.16.2.0/docs/Data-List-NonEmpty.html#t:NonEmpty)?

Reply:
(in MultiOr/MultiApp)

Reply:
The only problem is that I need to call the field a different name then :shrug: (conflict with `App` constructor field type)